ext/lib/mgmt/mcumgr/samples/smp_svr/zephyr/src/main.c

```diff
@@ -0,0 +1,221 @@
+
+/* Assign a name to the `ticks` stat. */
+STATS_NAME_START(smp_svr_stats)
+    STATS_NAME(smp_svr_stats, ticks)
+STATS_NAME_END(smp_svr_stats)
+
+/* Define an instance of the stats group. */
+STATS_SECT_DECL(smp_svr_stats) smp_svr_stats;
+
+struct fcb smp_svr_fcb;
+struct mdlog smp_svr_log;
+
+static const struct bt_data ad[] = {
+    BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+    BT_DATA_BYTES(BT_DATA_UUID128_ALL,
+        0x84, 0xaa, 0x60, 0x74, 0x52, 0x8a, 0x8b, 0x86,
+        0xd3, 0x4c, 0xb7, 0x1d, 0x1d, 0xdc, 0x53, 0x8d),
+};
+
+static const struct bt_data sd[] = {
+    BT_DATA(BT_DATA_NAME_COMPLETE, DEVICE_NAME, DEVICE_NAME_LEN),
+};
+
+static void advertise(void)
+{
+    int rc;
+
+    bt_le_adv_stop();
+
+    rc = bt_le_adv_start(BT_LE_ADV_CONN, ad, ARRAY_SIZE(ad),
+                         sd, ARRAY_SIZE(sd));
+    if (rc) {
+        printk("Advertising failed to start (rc %d)\n", rc);
+        return;
+    }
+
+    printk("Advertising successfully started\n");
+}
+
+static void connected(struct bt_conn *conn, u8_t err)
+{
+    if (err) {
+        printk("Connection failed (err %u)\n", err);
+    } else {
+        printk("Connected\n");
+    }
+}
+
+static void disconnected(struct bt_conn *conn, u8_t reason)
+{
+    printk("Disconnected (reason %u)\n", reason);
+    advertise();
+}
+
+static struct bt_conn_cb conn_callbacks = {
+    .connected = connected,
+    .disconnected = disconnected,
+};
+
+static void bt_ready(int err)
+{
+    if (err) {
+        printk("Bluetooth init failed (err %d)\n", err);
+        return;
+    }
+
+    printk("Bluetooth initialized\n");
+
+    advertise();
+}
+
+static int init_fcb(void)
+{
+    /* If the user has enabled FCB support, assume there is an FCB in the NFFS
+     * area.
+     */
+#ifndef CONFIG_FCB
+    return 0;
+#endif
+
+    static struct flash_sector sectors[128];
+
+    uint32_t sector_cnt;
+    int rc;
+
+    sector_cnt = sizeof sectors / sizeof sectors[0];
+    rc = flash_area_get_sectors(4, &sector_cnt, sectors);
+    if (rc != 0) {
+        return rc;
+    }
+
+    if (sector_cnt <= 1) {
+        return -EINVAL;
+    }
+
+    /* Initialize the FCB in the NFFS flash area (area 4) with one scratch
+     * sector.
+     */
+    smp_svr_fcb = (struct fcb) {
+        .f_magic = 0x12345678,
+        .f_version = MDLOG_VERSION,
+        .f_sector_cnt = sector_cnt - 1,
+        .f_scratch_cnt = 1,
+        .f_sectors = sectors,
+    };
+
+    rc = fcb_init(4, &smp_svr_fcb);
+    if (rc != 0) {
+        return rc;
+    }
+
+    return 0;
+}
+
+void main(void)
+{
+    int rc;
+
+    rc = init_fcb();
+    if (rc != 0) {
+        printk("Failed to initialize FCB; rc=%d\n", rc);
+    }
+
+#ifdef CONFIG_REBOOT_LOG
+    reboot_log_configure(&smp_svr_log);
+#endif
+
+    rc = STATS_INIT_AND_REG(smp_svr_stats, STATS_SIZE_32, "smp_svr_stats");
+    assert(rc == 0);
+
+    /* Register the built-in mcumgr command handlers. */
+#ifdef CONFIG_MCUMGR_CMD_FS_MGMT
```

Comment:
It is considered to be good-practice to do

if(IS_ENABLED(CONFIG_MCUMGR_CMD_FS_MGMT)) {

}

instead.

Reply:
`IS_ENABLED()` seems to elicit some compiler warnings if the relevant symbol isn't defined (e.g., if `CONFIG_MCUMGR_CMD_FS_MGMT` is not enabled):
```
/Users/ccollins/repos/mynewt-mcumgr/samples/smp_svr/zephyr/src/main.c: In function 'main':
/Users/ccollins/repos/mynewt-mcumgr/samples/smp_svr/zephyr/src/main.c:115:9: warning: implicit declaration of function 'fs_mgmt_register_group' [-Wimplicit-function-declaration]
         fs_mgmt_register_group();
         ^~~~~~~~~~~~~~~~~~~~~~
```

Reply:
That makes sense, my mistake.